planner/core/expression_rewriter.go

```diff
@@ -943,10 +943,38 @@ func (er *expressionRewriter) isNullToExpression(v *ast.IsNullExpr) {
 }
 
 func (er *expressionRewriter) positionToScalarFunc(v *ast.PositionExpr) {
-	if v.N > 0 && v.N <= er.schema.Len() {
-		er.ctxStack = append(er.ctxStack, er.schema.Columns[v.N-1])
-	} else {
-		er.err = ErrUnknownColumn.GenWithStackByArgs(strconv.Itoa(v.N), clauseMsg[er.b.curClause])
+	stkLen := len(er.ctxStack)
+	if val, ok := v.N.(*driver.ValueExpr); ok {
+		pos, isPos := val.GetValue().(int64)
+		if isPos && pos > 0 && pos <= int64(er.schema.Len()) {
+			er.ctxStack = er.ctxStack[:stkLen-1]
+			er.ctxStack = append(er.ctxStack, er.schema.Columns[pos-1])
+			return
+		}
+		er.err = ErrUnknownColumn.GenWithStackByArgs(strconv.Itoa(int(pos)), clauseMsg[er.b.curClause])
+	}
+	if val, ok := v.N.(*driver.ParamMarkerExpr); ok {
+		if len(er.ctx.GetSessionVars().PreparedParams) == 0 {
+			return
+		}
+		if param, err := expression.GetParamExpression(er.ctx, val, er.useCache()); err == nil {
```

Comment:
`GetParamExpression` has already been called in `expressionRewriter::Leave` for `ParamMarkerExpr`? so there should be a `Constant` on `er.CtxStack` that we can pop out and use here, thus we don't need to call `GetParamExpression` again?
I did no experimental verification since this branch cannot compile on my laptop due to the parser change.

Reply:
Thank you for the detailed comment. The updated PR should use ``Constant`` on ``er.ctxStack``.